### PR TITLE
fix: validate project name to be alphanumeric

### PR DESCRIPTION
**Pull request by aeng**
https://zanata.atlassian.net/browse/ZNTA-197

JIRA issue URL: *paste URL here*

*[Add a short description of what is changed here.]*

## Reviewer tasks

The following is based on the
[Zanata Development Guidelines](https://github.com/zanata/zanata-platform/wiki/Development-Guidelines).
As a reviewer, you should check the guidelines regularly for updates, and just
to refresh your memory.

The reviewer can use this list for reference to make sure
they have considered all the important points.

Use the checkboxes or not, whatever works best for you.

 - Code quality
   - [ ] Code passes style check (checkstyle/eslint)
   - [ ] Code is clear and concise
   - [ ] UI code is internationalised
   - [ ] Code has adequate comments where appropriate
   - Code is in an appropriate place
     - [ ] Classes are in appropriate packages
     - [ ] Code fits with class's responsibilities (SRP)
   - [ ] Configuration files are documented enough
   - If code is removed
     - [ ] No remaining code references the removed code
       - [ ] No orphaned rewrite rules
       - [ ] No orphaned config settings
     - [ ] No remaining comments refer to the removed code
     - [ ] Unused imports and libraries are removed
 - Testing
   - [ ] New code has tests
   - [ ] Changed code has tests
   - [ ] All tests pass
   - [ ] Test coverage stable or increased
 - Documentation
   - [ ] New features are documented
   - [ ] Docs removed for deleted features
   - [ ] Docs updated for all changed features
   - [ ] Developer/sysadmin docs updated if appropriate
 - If there are new dependencies
   - [ ] Does each new dependency pass all the
         [Considerations for new dependencies](https://github.com/zanata/zanata-platform/wiki/Development-Guidelines#new-technologies-and-dependencies)?

----
*This template can be updated in .github/PULL_REQUEST_TEMPLATE.md*

---
This change is [Reviewable](https://reviewable.io/reviews/zanata/zanata-platform/428)

### CODE REVIEW

File: server/zanata-war/src/main/java/org/zanata/action/ProjectHome.java [hunk 1]
```diff
@@ -768,6 +785,11 @@ public String persist() {
             return null;
         }
         getInstance().setSlug(getInputSlugValue());
+        if (!isValidName(getInstance().getName())) {
+            facesMessages.addGlobal(SEVERITY_ERROR,
+                    "This project name is must be alphanumeric");
```
Comment (djansen-redhat) [inline server/zanata-war/src/main/java/org/zanata/action/ProjectHome.java]: "Project name must be alphanumeric" - also, probably should be a Messages object.

File: server/zanata-war/src/main/java/org/zanata/action/ProjectHome.java [hunk 2]
```diff
@@ -688,6 +688,20 @@ public void initialize() {
         }
     }
 
+    public void onProjectNameChange(ValueChangeEvent e) {
+        if (!isValidName((String) e.getNewValue())) {
+            String componentId = e.getComponent().getId();
+            facesMessages.addToControl(componentId,
+                    "This project name is must be alphanumeric");
```
Comment (djansen-redhat) [inline server/zanata-war/src/main/java/org/zanata/action/ProjectHome.java]: "Project name must be alphanumeric" - also, probably should be a Messages object.

File: server/zanata-war/src/main/java/org/zanata/action/ProjectHome.java [hunk 3]
```diff
@@ -740,6 +754,9 @@ public String update() {
                 && !getSlug().equals(getInputSlugValue())) {
             getInstance().setSlug(getInputSlugValue());
         }
+        if (!isValidName(getInstance().getName())) {
+            return null;
```
Comment (djansen-redhat) [inline server/zanata-war/src/main/java/org/zanata/action/ProjectHome.java]: No facesmessage here?

**Comment by aeng** [comment]
@djansen-redhat done.

**Comment by huangp** [comment]
Reviewed 1 of 2 files at r1, 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[server/zanata-war/src/main/java/org/zanata/action/ProjectHome.java, line 704 at r2](https://reviewable.io:443/reviews/zanata/zanata-platform/428#-KpSX3IKXASPyWbJpw-J:-KpSX3IKXASPyWbJpw-K:bjel7sq) ([raw file](https://github.com/zanata/zanata-platform/blob/6fea6d8d682ed8d629a5542fb04c783d5f007de7/server/zanata-war/src/main/java/org/zanata/action/ProjectHome.java#L704)):*
> ```Java
>      */
>     public boolean isValidName(String name) {
>         String trimmedName = StringUtils.deleteWhitespace(name);
> ```

This will also delete whitespace in the middle of the name. We always allow that and I think there are plenty of projects using that as name.

---

*[server/zanata-war/src/main/java/org/zanata/action/ProjectHome.java, line 706 at r2](https://reviewable.io:443/reviews/zanata/zanata-platform/428#-KpSXJOclezvRJNIkgEn:-KpSXJOclezvRJNIkgEo:b3vw3vb) ([raw file](https://github.com/zanata/zanata-platform/blob/6fea6d8d682ed8d629a5542fb04c783d5f007de7/server/zanata-war/src/main/java/org/zanata/action/ProjectHome.java#L706)):*
> ```Java
>         String trimmedName = StringUtils.deleteWhitespace(name);
>         for (char c : trimmedName.toCharArray()) {
>             if (Character.isDigit(c) || Character.isLetter(c)) {
> ```

A regex might be easier. But what about people wanting to use unicode name? I think we just want to stop people using punctuation only in the name?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/428)*

**Comment by aeng** [comment]
@huangp The method is just stripping the whitespace and return true if there is alphanumeric in the string. It does not remove the whitespace when saved. The `if (Character.isDigit(c) || Character.isLetter(c))` does allow unicode name.

**Comment by huangp** [comment]
:lgtm: -(Awaiting review) +Reviewed +(On QA) 

---

Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/428#-:-KpSpe88ghP4xghJLAlq:b49ms4)*

File: server/zanata-war/src/main/resources/messages.properties [hunk 4]
```diff
@@ -331,6 +331,7 @@ jsf.project.WebhookName=Webhook name
 jsf.project.WebhookType.label=Type
 jsf.project.InvalidUrl=Invalid URL: {0}
 jsf.project.DuplicateUrl=Same URL is already in the list. {0}
+jsf.project.name.validation.alphanumeric=This project name is must be alphanumeric
```
Comment (djansen-redhat) [inline server/zanata-war/src/main/resources/messages.properties]: ~~**This**~~ project name ~~**is**~~ must be alphanumeric

Comment (aeng) [inline server/zanata-war/src/main/resources/messages.properties]: Project name must contains at least one alphanumeric character?

**Comment by djansen-redhat** [comment]
Merge when green 👍

**Comment by aeng** [comment]
jenkins built is green. https://zanata-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/github-zanata-org/job/zanata-platform/view/Pull%20Requests/job/PR-428/7/

**Pull request merged**